src/Language/Haskell/TH/Datatype.hs

```diff
@@ -202,6 +202,9 @@ data DatatypeInfo = DatatypeInfo
   , datatypeVars      :: [TyVarBndrUnit]   -- ^ Type parameters
   , datatypeInstTypes :: [Type]            -- ^ Argument types
   , datatypeVariant   :: DatatypeVariant   -- ^ Extra information
+  , datatypeReturnKind:: Kind              -- ^ Return Kind of the type
+                                           --
+                                           -- Evaluates to an error before GHC-8.10/template-haskell-2.16.
```

Comment:
This ought to say something about what happens when `th-abstraction` cannot find any return kind. In particular, if someone where to call [`normalizeInfo`](https://hackage.haskell.org/package/th-abstraction-0.6.0.0/docs/Language-Haskell-TH-Datatype.html#v:normalizeInfo) with something like `[d| newtype T = MkT Int# |]`, then `th-abstraction` would have no way to know with certainty what the return kind is without something akin to full type inference, which `th-abstraction` does not do.

If I understand this code in your patch:

```hs
     kind <- case mKind of
       Just kind -> pure kind
       Nothing -> pure StarT `recover` reifyType name
```

Then `th-abstraction` will conservatively set the `datatypeReturnKind` to `StarT` when it cannot figure out what the return kind should be. This is fine, but please document this in the Haddocks.